### PR TITLE
Issue 238 javier

### PR DESCRIPTION
**Pull request by javerom**
Added functionality in `esp32serial.py` to remember the last command issued. 

- This was done by adding a function parameter attribute to the `ESP32Serial` class that is initialized to `None`, but on issuing a command, is set to a function pointer with the correct arguments. 
- `self.connection.write` is wrapped in a `self._write(cmd, func)` function that encodes the `cmd` and writes it to the ESP32. The second argument is the function pointer to be called on retry. 
- All instances (only 3! `get`, `set`, and `get_all`) of `self.connection.write` are replaced by `self._write`, and the `func` argument is just a `lambda` function pointer to the corresponding function.

### CODE REVIEW

**Comment by javerom** [comment]
- Only have critical alarm handler without retry
- Added text box for alarm details (maintenance info)

**Comment by javerom** [comment]
Also, removed all `MessageBox` from the project except for in `start_stop_worker` which requires firmware changes to get the correct alarm/warning

**Comment by javerom** [comment]
- All command repeats are now encapsulated in `ExceptionWrapper`
- Cleanup of error message display

File: gui/communication/esp32serial.py [hunk 1]
```diff
@@ -116,19 +132,17 @@ def set(self, name, value):
             # but I don't really remember now the version running on
             # Raspbian
             command = 'set ' + name + ' ' + str(value) + '\r\n'
-            self.connection.write(command.encode())
+            try:
+                self._write(command)
+            except ESP32Exception:
+                raise
```
Comment (marcodeltutto) [inline gui/communication/esp32serial.py:138]: There are many blocks like this one
```
try:
    [some code]
except ESP32Exception: raise
```
Do we need this? Or can we remove the `try/except`? 
The `ESP32Exception` is raised by the `_write`, `get`, or `get_all` methods. 
We want the `ESP32Exception` to be raised so it can be caught by the `ExceptionWrapper`.

File: gui/special/special.py [hunk 2]
```diff
@@ -67,7 +67,9 @@ def _get_lung_recruit_eta(self):
         """
         Retrieves the Lungh Recruitment ETA from the esp32 and displays the result in Stop button
         """
-        eta = float(self._esp32.get("pause_lg_time"))
+        try:
+            eta = float(self._esp32.get("pause_lg_time"))
+        except ESP32Exception: return
```
Comment (marcodeltutto) [inline gui/special/special.py:72]: Same comment as before

File: gui/alarm_handler.py [hunk 3]
```diff
@@ -92,15 +91,7 @@ def _on_click_snooze(self):
...
-        except ESP32Exception as error:
-            msg = MessageBox()
-            func = msg.critical("Critical",
-                                "Severe hardware communication error",
-                                str(error),
-                                "Communication error",
-                                {msg.Retry: lambda: None,
-                                 msg.Abort: lambda: None})
-            func()
+        except ESP32Exception: return
```
[the remark this answers is not in the document]
Comment (javerom) [reply gui/alarm_handler.py:94]: Basically what's happening here is the ESP32Exception has already been handled by the `ExceptionWrapper`, but the function that called the ESP32 function that has the exception needs to know what to do now that the function has failed (it is still running in the loop). The behavior I chose was for the `ExceptionWrapper` to pass along the exception and for the calling function to return prematurely (i.e. not execute further code). We could let the function just return without an Exception, but then each calling function would need to know about default values and safety for the code that is executed post ESP32 failure.

File: gui/special/special.py (hunk 2 above)
Comment (javerom) [reply gui/special/special.py:72]: This is the same as in `alarm_handler`. If you can think of a better way to handle calling functions exiting prematurely on error, let me know.

File: gui/communication/esp32serial.py (hunk 1 above)
Comment (javerom) [reply gui/communication/esp32serial.py:138]: Ah yes I think these cases can have the `try/except` removed since the exception is raised anyways. I think this is left over from when there were different kinds of generic exceptions floating around.

Comment (javerom) [reply gui/communication/esp32serial.py:138]: I'll remove these and test to be sure that functionality is still ok

File: gui/alarm_handler.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (marcodeltutto) [reply gui/alarm_handler.py:94]: Ah I see! I think that if we arrive to the situation where the `ExceptionWrapper` stills fails after trying several times, further execution of the program should be stopped, and we should have the RSOD. In this case the `ExceptionWrapper` would not raise another exception, but would just display the RSOD and stop further code execution (I think further code execution can be suspended by showing the RSOD with `.exec()` instead of `.show()`).
I am saying this because if we are in the situation in which the exception is thrown even after 10 trials, then we can't trust anything I believe, and the only way would be to power cycle the Pi, or call for maintenance. 
Or am I missing another scenario in which we would like to have the code running?

[the remark this answers is not in the document]
Comment (javerom) [reply gui/alarm_handler.py:94]: Yeah I see what you're saying. I think we need a way of entering RSOD without killing the program. I wonder what the behavior is if we switch to RSOD and stop the Qt event handler (i.e. all QTimers stop and the window freezes)? Maybe this is the behavior we want? I'll play around with that.

File: gui/alarm_handler.py [hunk 4]
```diff
@@ -319,26 +292,37 @@ def __init__(self, mainparent, esp32):
...
         self._label_criticalerror = mainparent.findChild(QtWidgets.QLabel, "label_criticalerror")
+        self._label_criticaldetails = mainparent.findChild(
+                QtWidgets.QLabel, 
+                "label_criticaldetails")
         self._button_retrycmd = mainparent.findChild(QtWidgets.QPushButton, "button_retrycmd")
 
         self._button_retrycmd.pressed.connect(self._retry_cmd)
 
-    def show_critical_error(self, text):
+    def show_critical_error(self, text, details=None):
```
Comment (carmelopellegrino) [inline gui/alarm_handler.py:302]: if the default value of `details` is `""` (empty string) there is no need to test for `is None`

File: gui/alarm_handler.py [hunk 5]
```diff
@@ -319,26 +292,37 @@ def __init__(self, mainparent, esp32):
...
         This includes changing the screen to red and displaying a big message to this effect.
         """
         self._label_criticalerror.setText(text)
         self._toppane.setCurrentWidget(self._criticalerrorpage)
         self._bottombar.setCurrentWidget(self._criticalerrorbar)
+        if details is not None:
+            self._label_criticaldetails.setText(details)
+
+        QtCore.QCoreApplication.processEvents()
+        QtCore.QCoreApplication.quit()
```
Comment (carmelopellegrino) [inline gui/alarm_handler.py:314]: does this close the application? (in the sense that the gui completely closes)?

File: gui/alarm_handler.py [hunk 6]
```diff
@@ -319,26 +292,37 @@ def __init__(self, mainparent, esp32):
...
 
-    def call_system_failure(self):
+    def call_system_failure(self, details=None):
```
Comment (carmelopellegrino) [inline gui/alarm_handler.py:316]: same applies here for `None` vs `""`

File: gui/mainwindow.ui [hunk 7]
```diff
@@ -730,6 +736,30 @@ Call the Maintenance Service</string>
           </property>
          </widget>
         </item>
+        <item>
+         <widget class="QLabel" name="label_criticaldetails">
+          <property name="font">
+           <font>
+            <family>Noto Mono</family>
```
Comment (carmelopellegrino) [inline gui/mainwindow.ui:743]: remember to add the `Noto` font to the list of fonts to install in the readme :)

File: gui/mvm_gui.py [hunk 8]
```diff
@@ -65,20 +66,32 @@ def main():
...
+
+    # Assign exception function
+    esp32.assign_except_func(window.critical_alarm_handler.call_system_failure)
+
+    # Set up watchdog and star the main Qt executable
+    esp32.set("wdenable", 1)
+    watchdog = QtCore.QTimer()
+    def watchdog_redirect():
+        try: esp32.set_watchdog()
+        except ESP32Exception: return
```
Comment (carmelopellegrino) [inline gui/mvm_gui.py:89]: will `esp32.set_watchdog()` ever raise a `ESP32Exception`? Shouldn't it be catched in the wrapper?

File: gui/alarm_handler.py (hunk 4 above)
Comment (javerom) [reply gui/alarm_handler.py:302]: Ok can change

File: gui/alarm_handler.py (hunk 5 above)
Comment (javerom) [reply gui/alarm_handler.py:314]: This closes the GUI and effective returns from `app.exec()` in `mvm_gui.py`. After that, I put a catch to pause the terminal in the case of an error state from the wrapper class.

File: gui/alarm_handler.py (hunk 6 above)
Comment (javerom) [reply gui/alarm_handler.py:316]: Done

File: gui/mvm_gui.py (hunk 8 above)
Comment (javerom) [reply gui/mvm_gui.py:89]: You are right. There was a problem with an unhandled exception in the `QTimer` loop before that caused a GUI crash when the watchdog would throw the exception, but this is no longer the case with the new halting of the Qt event handler. I will amend.

File: gui/mainwindow.ui (hunk 7 above)
Comment (javerom) [reply gui/mainwindow.ui:765]: We use `Noto Mono` for the monitors. Had we not added the font to the readme from that?

Comment (carmelopellegrino) [reply gui/mainwindow.ui:765]: Eheheh... no :)

File: gui/alarm_handler.py (hunk 5 above)
Comment (carmelopellegrino) [reply gui/alarm_handler.py:313]: OK, thanks for the confirmation! I think it is not the right thing to do here. The program should run forever while displaying the red error screen, which is not what I would expect if `app._exec()` returns.

Comment (javerom) [reply gui/alarm_handler.py:313]: You are right. I've been having issues with the event handler sometime leaking through QTimer functions, but if we stop it right after RSOD, nothing happens. I'll keep the `processEvents` since that allows the RSOD to show up instead of sometimes just freezing on the current view, but I'll replace the `quit()` with an `input()`, which will hang forever

**Comment by javerom** [comment]
Fixed many of the mini-issues mentioned above to streamline code. Additionally, on every retry, a reconnect attempt to ESP32 is made. Reconnect was tests by unbinding then quickly rebinding the port, which seems to allow the GUI to recover.

File: gui/exception_wrapper.py [hunk 9]
```diff
@@ -0,0 +1,105 @@
...
+                    try:
+                        self.except_func()
+                    except self.ExceptionType:
+                        pass
+                print(self._last_error_str)
+
+                
+        # Lock the state as failed by setting failed_state
+        if self.failed_func is not None and not self.failed_state:
+            self.failed_state = True
```
Comment (carmelopellegrino) [inline gui/exception_wrapper.py:55]: is the `failed_state` needed anymore?

Comment (javerom) [reply gui/exception_wrapper.py:55]: Good point. It was needed before when many functions could be calling the wrapper, but that is no longer the case. I'll remove.

**Comment by javerom** [comment]
Bticodes that cause RSOD are configurable via `default_settings.yaml`, which is currently set to just bit 31. All GUI functions freeze on these error code and can only be recovered by reboot.
